### PR TITLE
Use local L2 projection for coarsening L2 spaces on non-conforming AMR meshes [bugfix/rwa/coarsen-L2-proj]

### PR DESCRIPTION
**Pull request by rw-anderson**
When transfering a GridFunction from fine to coarse using a discontinuous space, use an element-wise L2 projection operator. This will be conservative and more accurate than a pointwise interpolation.

- [x] Add comprehensive unit tests (order, basis type, element type, etc.) for correctness of L2 projections in DG spaces.
- [x] Add stress test for parallel correctness
- [x] Try to collapse loop in fespace.cpp that does the global assembly for the DG and CG cases, a lot of duplication.
- [x] Same as above, but in pfespace.cpp
- [x] Support triangles and tets in non-conforming mode
- [x] Change tests to use basis change rather than exact projection
- [x] Revert exact projection code (will be added as an option in #3015)
- [x] look at map_type INTEGRAL case
- [x] Remove ex_test.cpp and ex_testp.cpp once MR is approved. Leaving in-place in case needed for any further debugging.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3007](https://github.com/mfem/mfem/pull/3007) | @rw-anderson | @tzanio | @tzanio + @jakubcerveny + @brendankeith + @camierjs + @psocratis | 5/24/22 | 1/21/23 | ⌛due 6/14/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by rw-anderson** [comment]
The parallel case turned out to be simple after all.

File: fem/fe/fe_pos.cpp
Comment (rw-anderson) [inline fem/fe/fe_pos.cpp:27]: This change is more or less orthogonal to the purpose of this PR. However, a lot of my testing was predicated on needing exact projections of initial data using the Bernstein basis. But now I am a little bit stuck. If I remove this and address projection in a separate PR, the tests will fail. If I merge as-is, we at least temporarily lose the ability to use the original monotone interpolation operator. Another option is to merge both PR's into this one: the L2 proj for coarsening and the L2 projection for projection of coefficients.

File: mesh/mesh.cpp
Comment (rw-anderson) [inline mesh/mesh.cpp:9088]: This is a somewhat orthogonal bugfix. When using op=0, the minimum operator for aggregation, initializing to 0 means you never get anything but 0 as the final result (assuming you aren't supplying negative errors).

File: mesh/ncmesh.cpp
Comment (rw-anderson) [inline mesh/ncmesh.cpp:4288]: This was also an orthogonal bugfix, this clause is needed for 1D non-conforming refinement.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/fe/fe_pos.cpp
Comment (tzanio) [inline fem/fe/fe_pos.cpp:1]: Lets reverth this 😄

**Comment by rw-anderson** [comment]
I did a little testing with Laghos.

I changed the line:
L2_FECollection L2FEC(order_e, dim/*, BasisType::Positive*/);

to use the positive basis, as we do in non-AMR Laghos (and BLAST).

Then ran:
mpirun -np 8 laghos -p 1 -m ../data/square01_quad.mesh -rs 4 -tf 0.8 -amr -vis

* crashes with develop mfem.
* runs to completion with this branch:
density-amr-grid
density-amr-nogrid

**Comment by artv3** [comment]
> I did a little testing with Laghos.
> 
> I changed the line: L2_FECollection L2FEC(order_e, dim/_, BasisType::Positive_/);
> 
> to use the positive basis, as we do in non-AMR Laghos.
> 
> Then ran: mpirun -np 8 laghos -p 1 -m ../data/square01_quad.mesh -rs 4 -tf 0.8 -amr -vis
> 
>     * crashes with develop mfem.
> 
>     * runs to completion with this branch:
>       density-amr-grid
>       density-amr-nogrid

Nice! would it work with partial assembly? -pa?

**Comment by rw-anderson** [comment]
> > I did a little testing with Laghos.
> > I changed the line: L2_FECollection L2FEC(order_e, dim/_, BasisType::Positive_/);
> > to use the positive basis, as we do in non-AMR Laghos.
> > Then ran: mpirun -np 8 laghos -p 1 -m ../data/square01_quad.mesh -rs 4 -tf 0.8 -amr -vis
> > ```
> > * crashes with develop mfem.
> > 
> > * runs to completion with this branch:
> >   density-amr-grid
> >   density-amr-nogrid
> > ```
> 
> Nice! would it work with partial assembly? -pa?

I don't know. The docs for Laghos say:

"Partial assembly is currently not supported in AMR mode."
and it seems to automatically turn that option off if it's on in the AMR variant.

I'm not sure if that's outdated, though, and if there's any reason it couldn't be supported now with more recent work in MFEM.

**Comment by tzanio** [comment]
ping: @camierjs

**Comment by camierjs** [comment]
The [amr branch](https://github.com/ceed/Laghos/tree/amr) of Laghos does work with partial assembly.
I will try it again with these  changes.

File: mesh/mesh.cpp
Comment (v-dobrev) [reply mesh/mesh.cpp:9088]: This is now incorrect for the case `op==1` since the loop below starts from 0.

Comment (rw-anderson) [reply mesh/mesh.cpp:9088]: I think this is a github bug or limitation. The loop starts from 1 now.

**Comment by camierjs** [comment]
Tested successfully on the [amr](https://github.com/CEED/Remhos/tree/amr) branch of Remhos:
| master |  coarsen-L2-proj |
|---|---|
| image | image |
|Mass loss u:   6.10436e-05 | Mass loss u:   2.33147e-14|

`mpirun -n 6 ./remhos 3 -rs 1 -tf 1 -vs 10 -o 3 -amr -ae 2 -df 1.0 -ar 0.95 -ad 0.98 -am 4`

**Comment by artv3** [comment]
@rw-anderson out of curiosity could 

> Tested successfully on the [amr](https://github.com/CEED/Remhos/tree/amr) branch of Remhos:
> master 	coarsen-L2-proj
> image 	image
> Mass loss u: 6.10436e-05 	Mass loss u: 2.33147e-14
> 
> `mpirun -n 6 ./remhos 3 -rs 1 -tf 1 -vs 10 -o 3 -amr -ae 2 -df 1.0 -ar 0.95 -ad 0.98 -am 4`

@rw-anderson when we do the coarsening would it still preserve bounds?

**Review by psocratis — COMMENTED** [review]
This looks good to me so far! Are we waiting for the pending TODO items to push it through?

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by rw-anderson** [comment]
> @rw-anderson when we do the coarsening would it still preserve bounds? In regards to the Remhos example.

No, simple L2 projection will definitely not be bounds preserving. So this is something to think about further.

**Comment by rw-anderson** [comment]
I think this should be ready now (long last).

**Comment by tzanio** [comment]
Thanks @rw-anderson 🎉 

@camierjs, @jakubcerveny, @psocratis, and @brendankeith -- can you please take another look when you get a chance?

**Comment by artv3** [comment]
> > @rw-anderson when we do the coarsening would it still preserve bounds? In regards to the Remhos example.
> 
> No, simple L2 projection will definitely not be bounds preserving. So this is something to think about further.

One option is to use the ElementFCTProjection method in Remhos, we used it over the summer in lieu of L2 projection for a similar use case.

**Comment by camierjs** [comment]
@rw-anderson, do we still need `examples/ex_test[p].cpp` or is `tests/unit/fem/test_derefine.cpp` covering the tests you wanted? We can also move it inside the `test_derefine`, accessible with the `--all` option.

**Comment by rw-anderson** [comment]
> @rw-anderson, do we still need `examples/ex_test[p].cpp` or is `tests/unit/fem/test_derefine.cpp` covering the tests you wanted? We can also move it inside the `test_derefine`, accessible with the `--all` option.

These were just for playing around while I was developing this - test worth keeping were already moved into test_derefine. I've removed the files.

**Comment by tzanio** [comment]
@jakubcerveny, this is ready to be merged. Do you have any questions or concerns?

File: fem/fespace.cpp
Comment (psocratis) [inline fem/fespace.cpp:2096]: How is this different than using `SetRow` ?

**Review by psocratis — APPROVED** [review]

File: fem/fespace.cpp
Comment (v-dobrev) [reply fem/fespace.cpp:2096]: The complexity of `SetRow` is linear in the number of nonzeroes in the row. The complexity of the loop using `Set(int,int,double)` is quadratic, so `SetRow` should be preferred.

**Comment by camierjs** [comment]
[Laghos](https://github.com/CEED/Laghos) looks good too!

image

`mpirun -n 3 ./laghos -p 3 -m data/rectangle01_quad.mesh -pa -vis -vw 1 -vk mAcRjlppppppp -ok 3 -ot 2 -rs 1 -d cpu -amr -ae 1 -am 5 -aj 0.75 -ad 0.1`
Awesome @rw-anderson !

**Review by camierjs — APPROVED** [review]

**Review by jakubcerveny — APPROVED** [review]
Nice, I'm just not sure about losing the `SetRow` optimization.

File: fem/fespace.cpp
Comment (tzanio) [reply fem/fespace.cpp:2096]: @rw-anderson is there a reason while `SetRow` was replaced with the explicit loop? Can we revert?

**Comment by tzanio** [comment]
@rw-anderson, we are ready to merge this after the `SetRow` comment is addressed. Can you take a look when you get a chance?

**Comment by tzanio** [comment]
@psocratis, @v-dobrev and @jakubcerveny -- the `SetRow` change has been added now. Can you take a final look before we merge in `next`?

**Comment by psocratis** [comment]
> @psocratis, @v-dobrev and @jakubcerveny -- the `SetRow` change has been added now. Can you take a final look before we merge in `next`?

looks good to me

**Comment by v-dobrev** [comment]
> @psocratis, @v-dobrev and @jakubcerveny -- the `SetRow` change has been added now. Can you take a final look before we merge in `next`?

I'm not a one of the assigned reviewers, I just commented about `SetRow`. I cannot comment about the PR overall.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks for this important capability improvement, @rw-anderson! 🚀 

@v-dobrev, @jakubcerveny, @psocratis and @camierjs -- _Thanks for reviewing!_

**Pull request merged**
